### ISSUE TITLE
Per-user access

### ISSUE THREAD

**Issue by Bogdan107** [firewalld/firewalld#725]
I am using a firewall on my laptop and I have no relationship to networking.

Production routers:
- have fixed set of software and not focused to controll activity of programs, which runned inside the router system;
- focused on serving external, forwarded network packets.

Personal laptop:
- have high probability to run unpriviledged programs at any time and admin need to controll programs activity too;
- mostly need to control input/output network packets, not forwarded packets.

Long time I used self-writted script for control nftables daemon.
Now I am learning firewald and have some troubles due to the fact that firewalld does not have some feature, because focused to serving forwarded network packets as many others firewall's...

I usу DENY policy by default, and allow network usage only for controlled list of system users.
Users with uid>=1000 can connect to internet only over local proxy (like squid or privoxy), and each user program (web-browser, torrent, media-player) for network activity outside the laptop must have their own login/password for proxy authentification. System programs (like ping, dig, whois) started under sudo with password.

Some lines from my script:
> \# Prepare sets:
> nft add rule inet filter output tcp  dport @tcp_port_out_allow  ct state new  counter accept;
> nft add rule inet filter input  tcp  dport @tcp_port_in_allow   ct state new  counter accept;
> nft add rule inet filter output udp  dport @udp_port_out_allow  ct state new  counter accept;
> nft add rule inet filter input  udp  dport @udp_port_in_allow   ct state new  counter accept;
> nft add rule inet filter output meta skuid @user_out_allow      ct state new  counter accept;
> nft add rule inet filter input  meta skuid @user_in_allow       ct state new  counter accept;
> 
> nft add rule inet filter output tcp  dport @tcp_port_out_deny  ct state new  counter drop;
> nft add rule inet filter input  tcp  dport @tcp_port_in_deny   ct state new  counter drop;
> nft add rule inet filter output udp  dport @udp_port_out_deny  ct state new  counter drop;
> nft add rule inet filter input  udp  dport @udp_port_in_deny   ct state new  counter drop;
> nft add rule inet filter output meta skuid @user_out_deny      ct state new  counter drop;
> nft add rule inet filter input  meta skuid @user_in_deny       ct state new  counter drop;
> 
> \# Allow basic network activity for all users:
> nft add element  inet filter  udp_port_out_allow { 53, 123, 1900 };
> nft add element  inet filter  tcpp_port_out_allow { 53, 123, 1900, 6667 };
>
> \# Allow full network activity for list of users:
> nft add element  inet filter  user_out_allow { root, portage, sync, ddclient, ntp, dnscrypt-proxy, miredo, i2pd, tor, privoxy };
> 
> \# Allow limited network activity for 'squid' user:
> nft add rule inet filter output meta skuid { squid } tcp  dport { 21, 80-88, 443 } ct state new  counter accept;

I need command line options for:
- control network usage per-user (list of users);
- control network usage per-port (list of ports/diapazon of ports) per-user (list of users).

**Comment by erig0** [comment]
What's missing is `skuid` and `skgid` matching. Then you could use rich rules.
e.g.
```
# firewalld -cmd --add-rich-rule='port port=1234 protocol=tcp socket uid=1000' accept
```

Firewalld currently lacks the `socked uid=1000` support.

**Comment by ragnvaldf** [comment]
I also need skuid and skgid matching on puppet-managed RHEL with firewalld+nftables. Is there any known workarounds for this use case?

**Comment by erig0** [comment]
> I also need skuid and skgid matching on puppet-managed RHEL with firewalld+nftables. Is there any known workarounds for this use case?

No. If you want to DROP traffic based on skuid/skgid, then you can use a direct rule. But ACCEPT won't work as the packet still needs to pass through nftables.

**Comment by ragnvaldf** [comment]
Thank you for advice @erig0 . This solved my current problem:
`firewall-cmd --direct --add-rule ipv4 filter OUTPUT 0 -d xxx.xxx.xxx.xxx/32 -p tcp -m tcp --dport 443 -m owner ! --uid only_allowed_user -j REJECT --reject-with icmp-port-unreachable`

**Comment by Talkless** [comment]
+1 for skuid/skgid matching support!

**Comment by erig0** [comment]
PSA: upstream nftables is working on INPUT and OUTPUT socket matching. Currently skuid/skgid matching is limited to OUTPUT only.

INPUT support means we could say things like "only apache can accept https connections".

**Comment by Bogdan107** [comment]
> PSA: upstream nftables has proposed patches for INPUT and OUTPUT socket matching. Currently skuid/skgid matching is limited to OUTPUT only.
> 
> INPUT support means we could say things like "only apache can accept https connections".

Yes! I want a quick and easy mechanism to limit just OUTPUT traffic!
This is an extremely high priority function for me!

With thit function, I can allow only WHITELISTED programs. So, any user installed programs, which needs to use internet, must be configured to use system proxy server (in my case - this is Privoxy) or configured personally by firewalld administrator.

## For example:

### Rule 1: Allow OUTPUT connections for whitelisted users:
- system users: **root**;
- dns servers: **unbound**, **dnsmasq**;
- proxy servers: **privoxy**, **squid**;
- overlay networks: **miredo**, **tor**, **i2p**, **yggdrasil**.

### Rule 2: Allow OUTPUT to the **internal** proxy servers:
- allow OUTPUT from localhost to 127.0.0.1:3128 (allow access to the local proxy-server Squid);
- allow OUTPUT from localhost to 127.0.0.1:8118 (allow access to the local proxy-server Privoxy);

### Rule 3: Allow OUTPUT to the **external** proxy servers:
- allow OUTPUT to <ip>/<port>

### Rule 4: Disallow OUTPUT connections for non-whitelisted users or non-configured programs:
- **LIMIT RATE** to the some very low value (for example - 8 KB/s);
- or deny OUTPUT ALL (to disallow unconfigured connections) (all allowed traffic is configured in rules 1..3).

## With this rules:
- all system services allowed to use internet connections;
- all un-whitelisted users can acces only to _local proxy server_, like Privoxy/Squid/etc (Rule 2) or to the _external proxy server_ (Rule3);
- any non-system program, which need to use internet connection, MUST have his own login/pass pairs, configured in proxy-server config;
- any programs, which does not have his own login/pass pair on the proxy server, can use internet only in a very low speed mode (Rule 4) or may be disabled by few clicks (disable LIMIT rule and allow DENY rule).

**Comment by Talkless** [comment]
I might be really mistaken (as I am not yet firewalld user), but that example `firewall-cmd --zone apache --add-service https
` would not work because `skuid`/`skgid` only works for `OUTPUT`, meanwhile zones in firewalld is about `INPUT`?

For _policies_ though I do get it, as policies implement OUTPUT / FORWARD filtering, right?

**Comment by erig0** [comment]
> I might be really mistaken (as I am not yet firewalld user), but that example `firewall-cmd --zone apache --add-service https ` would not work because `skuid`/`skgid` only works for `OUTPUT`, meanwhile zones in firewalld is about `INPUT`?

IIRC, there were recent nftables patches to allow INPUT socket matching. i.e. `meta skuid`, `meta skgid`, and `meta cgroup` should work on both INPUT/OUTPUT on modern kernel/nftables.

I have not verified this though.

**Comment by erig0** [comment]
Adding "blocked" because this requires support in nftables.
